<a href="https://colab.research.google.com/github/EmilMaliev/Database/blob/master/Project_SQL_%D0%9D%D0%BE%D1%83%D1%82%D0%B1%D1%83%D0%BA_%D1%88%D0%B0%D0%B1%D0%BB%D0%BE%D0%BD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [6]:
import pandas as pd
import psycopg2

In [11]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python

In [12]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [ ]:
# текст запроса
query_3_1 = f'''select COUNT(*) from vacancies
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

<ipython-input-30-7f364a65e0f6>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197


2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [ ]:
# текст запроса
query_3_2 = f'''select COUNT(*) from employers
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

<ipython-input-32-e0e2d23c59c3>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [ ]:
# текст запроса
query_3_3 = f'''select COUNT(*) from areas
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

<ipython-input-34-11c9948ba1f3>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [ ]:
# текст запроса
query_3_4 = f'''select COUNT(*) from industries
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

<ipython-input-36-aae953b7cc87>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

In [ ]:
# выводы по предварительному анализу данных

Можно заметить, что количтесво вакансий более чем 2 раза больше, чем работадателй. Это значит, что рабодатели нуждаются в нескольких специалистов на разные позции, сферы деятельности и т.п.

## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
query_4_1 = f'''select a.name area,
                       COUNT(*) cnt
                from areas a
                join vacancies v on v.area_id = a.id
                group by 1
                order by 2 DESC
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

<ipython-input-41-c5d40976a413>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = f'''select COUNT(*)
                from vacancies
                where salary_to is not null or salary_from is not null
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

<ipython-input-43-bebb05645b92>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [ ]:
# текст запроса
query_4_3 = f'''select round(avg(salary_from)),
                       round(avg(salary_to)) from vacancies
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

<ipython-input-47-3cb76ea8c081>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


round     round
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query_4_4 = f'''select schedule,
                       employment,
                       COUNT(*)
                from vacancies
                group by 1, 2
                order by 3 DESC
'''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

<ipython-input-51-ddd5034b1d4e>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [ ]:
# текст запроса
query_4_5 = f'''select experience,
                       count(*)
                from vacancies
                group by 1
                order by 2
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

<ipython-input-55-868d7ae2deeb>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

In [ ]:
# выводы по детальному анализу вакансий

По результатом запроса видно, что более крупные крупные города нуждаются в большем количестве специалистов, в основном на полную занятость и имеющие хотя бы от 1 года опыта работы.

## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# текст запроса
query_5_1 = f'''  (select e.name,
                          count(*)
                  from employers e
                  join vacancies v on v.employer_id = e.id
                  group by 1
                  order by 2 DESC
                  limit 1)
                  union
                  (select e.name,
                          count(*)
                  from employers e
                  join vacancies v on v.employer_id = e.id
                  group by 1
                  order by 2 DESC
                  offset 4 limit 1)
                  order by 2 DESC
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

<ipython-input-83-ad08e28dab7b>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
query_5_2 = f'''select  a.name region_name,
                        count(distinct e.id) employers_count
                from areas a
                left join employers e on a.id = e.area
                left join vacancies v on a.id = v.area_id
                where v.id is null
                group by a.name
                order by employers_count desc
                limit 1
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

<ipython-input-9-5d775e2f9142>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


region_name  employers_count
0      Россия              410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query_5_3 = f'''select e.name, count(distinct area_id)
                from employers e
                join vacancies v on v.employer_id = e.id
                group by 1
                order by 2 desc
                '''

In [ ]:
#результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

<ipython-input-172-6b9e1d533ed2>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [ ]:
# текст запроса
query_5_4 = f'''with employers_without_industries as (
                select id
                from employers
                except
                select employer_id
                from employers_industries
                )
                select count(*)
                from employers_without_industries
                '''


In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

<ipython-input-162-ef2fed5e214c>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [ ]:
# текст запроса
query_5_5 = f'''select e.name
            from employers e
            join employers_industries ei on ei.employer_id = e.id
            group by 1
            having count(industry_id) = 4
            order by 1
            offset 2
            limit 1
'''


In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

<ipython-input-165-a41bab2be398>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [ ]:
# текст запроса
query_5_6 = f'''select count(employer_id)
                from employers_industries ei
                join industries i on i.id = ei.industry_id
                where name = 'Разработка программного обеспечения'
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

<ipython-input-168-8e05d7ddda4a>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [ ]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)

page = BeautifulSoup(response.text, 'html.parser')

table = page.find('table', class_ = "standard sortable")
cities = []
if table:
  rows = table.find_all('tr')
  for row in rows[1:]:
        cols = row.find_all("td")
        if cols:
            city = cols[1].text
            cities.append(city)
cities

['Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Самара',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград']

In [ ]:
# текст запроса
query_5_7 = f'''with yandex_vacancies as(
                  select a.name city_name,
                  count(v.id) vacancy_cnt
                  from areas a
                  join vacancies v on v.area_id = a.id
                  join employers e on e.id = v.employer_id
                  where e.name = 'Яндекс' and a.name in {tuple(cities)}
                  group by 1
                )
                select *
                from yandex_vacancies
                union all
                select 'total' city_name,
                       sum(vacancy_cnt) vacancy_cnt
                from yandex_vacancies
                order by 2
                    '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

<ipython-input-32-df759ebf7dc8>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


city_name  vacancy_cnt
0              Омск         21.0
1         Челябинск         22.0
2        Красноярск         23.0
3         Волгоград         24.0
4             Пермь         25.0
5            Казань         25.0
6    Ростов-на-Дону         25.0
7            Самара         26.0
8               Уфа         26.0
9         Краснодар         30.0
10          Воронеж         32.0
11      Новосибирск         35.0
12  Нижний Новгород         36.0
13     Екатеринбург         39.0
14  Санкт-Петербург         42.0
15           Москва         54.0
16            total        485.0

***

In [ ]:
# выводы по анализу работодателей

По результатом запросов можно увидеть, что Яндекс имеет наибольшее количество вакансий и ищет работников в наибольшем количестве регионов, большая часть, которых рассредоточена на крупные города.

## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [ ]:
# текст запроса
query_6_1 = f'''select count(v.name)
                from vacancies v
                where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

<ipython-input-14-631a6abf7434>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [ ]:
# текст запроса
query_6_2 = f'''select count(name)
                from vacancies
                where (lower(name) like '%data science%'
                      or lower(name) like '%data scientist%'
                      or lower(name) like '%исследователь данных%'
                      or (lower(name) like '%ml%' and lower(name) not like '%html%')
                      or lower(name) like '%machine learning%'
                      or lower(name) like '%машинн%обучен%')
                      and (lower(name) like '%junior%' or experience = 'Нет опыта' or schedule = 'Стажировка')
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

<ipython-input-46-d5ce39ec296d>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query_6_3 = f'''select count(name)
                from vacancies
                where (lower(name) like '%data science%'
                      or lower(name) like '%data scientist%'
                      or lower(name) like '%исследователь данных%'
                      or (lower(name) like '%ml%' and lower(name) not like '%html%')
                      or lower(name) like '%machine learning%'
                      or lower(name) like '%машинн%обучен%')
                      and (lower(key_skills) like '%sql%' or lower(key_skills) like '%postgres%')
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

<ipython-input-53-93be484d31ff>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query_6_4 = f'''select count(name)
                from vacancies
                where (lower(name) like '%data science%'
                      or lower(name) like '%data scientist%'
                      or lower(name) like '%исследователь данных%'
                      or (lower(name) like '%ml%' and lower(name) not like '%html%')
                      or lower(name) like '%machine learning%'
                      or lower(name) like '%машинн%обучен%')
                      and (lower(key_skills) like '%python%')
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

<ipython-input-56-a429304bcb54>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [ ]:
# текст запроса
query_6_5 = f'''select round(avg(length(key_skills) - length(replace(key_skills, '\t', '')) + 1), 2)
                from vacancies
                where (lower(name) like '%data science%'
                      or lower(name) like '%data scientist%'
                      or lower(name) like '%исследователь данных%'
                      or (name like '%ML%' and lower(name) not like '%html%')
                      or lower(name) like '%machine learning%'
                      or lower(name) like '%машинн%обучен%')
                      and key_skills is not null and key_skills != ''
                '''


In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

<ipython-input-75-99a1fdc5a2fc>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [ ]:
# текст запроса
query_6_6 = f'''select experience,
                       round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from)) / 2))
                from vacancies
                where (lower(name) like '%data science%'
                      or lower(name) like '%data scientist%'
                      or lower(name) like '%исследователь данных%'
                      or (name like '%ML%' and lower(name) not like '%html%')
                      or lower(name) like '%machine learning%'
                      or lower(name) like '%машинн%обучен%')
                      and (salary_from is not null or salary_to is not null)
                group by 1
                '''

In [ ]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

<ipython-input-94-c2f1a4577ae9>:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

In [ ]:
# выводы по предметному анализу

По результатом запросов можно заметить, что количество вакансий для начинающих DS небольшое(51) и в среднем их зп составлет 74643. Можно сказать, что конкурс на место будет достаточно большим, особенно для вакансий, в которых предоставляются достаточно хороший условия для работы.

## Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

Дополнительные исследование

Сколько всего DS вакансий?

In [14]:
query_7_1 = f'''select count(name)
                from vacancies
                where (lower(name) like '%data science%'
                      or lower(name) like '%data scientist%'
                      or lower(name) like '%исследователь данных%'
                      or (lower(name) like '%ml%' and lower(name) not like '%html%')
                      or lower(name) like '%machine learning%'
                      or lower(name) like '%машинн%обучен%')
                      '''
df = pd.read_sql_query(query_7_1, connection)
df

<ipython-input-14-68b447e4747a>:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


count
0    536

По результатом запросов, можно сказать, что вакансий для начинающих DS небольшое количество, соответсвенно возможно конкуренция будет большая, особенно на более известные компании, такие как Яндекс. Стоит подметить, что и большая часть вакансий приходится на крупные города, поэтому если Вы проиживатете в более мелких городах, возможно будет отсуствовать вакансии для вашего города, следовательно надо рассчитывать на переезд или удаленную работу. Также можно увидеть, что чуть менее половины вакансий DS требует знания SQL и больше половины знания Python. Стоит это учесть, для прохождения собеседований, возможно лучше сделать упор побольше на эти навыки для своего портфолио.